From: Jan Beulich <jbeulich@suse.com>
Date: Tue, 19 Jun 2018 15:11:44 +0100
X-Dgit-Generated: 4.8.3+xsa267+shim4.10.1+xsa267-1+deb9u9~ 0a0aa230554115958fa7dc95ae922bf508d40eea
Subject: x86/mm: don't bypass preemption checks

While unlikely, it is not impossible for a multi-vCPU guest to leverage
bypasses of preemption checks to drive Xen into an unbounded loop.

This is XSA-264.

Signed-off-by: Jan Beulich <jbeulich@suse.com>
Reviewed-by: Andrew Cooper <andrew.cooper3@citrix.com>

---

--- xen-4.8.3+xsa267+shim4.10.1+xsa267.orig/xen/arch/x86/mm.c
+++ xen-4.8.3+xsa267+shim4.10.1+xsa267/xen/arch/x86/mm.c
@@ -2661,7 +2661,7 @@ static int _put_page_type(struct page_in
                 nx = x & ~(PGT_validated|PGT_partial);
                 if ( unlikely((y = cmpxchg(&page->u.inuse.type_info,
                                            x, nx)) != x) )
-                    continue;
+                    goto maybe_preempt;
                 /* We cleared the 'valid bit' so we do the clean up. */
                 rc = _put_final_page_type(page, x, preemptible, ptpg);
                 ptpg = NULL;
@@ -2697,12 +2697,13 @@ static int _put_page_type(struct page_in
              */
             cpu_relax();
             y = page->u.inuse.type_info;
-            continue;
+            goto maybe_preempt;
         }
 
         if ( likely((y = cmpxchg(&page->u.inuse.type_info, x, nx)) == x) )
             break;
 
+    maybe_preempt:
         if ( preemptible && hypercall_preempt_check() )
             return -EINTR;
     }
@@ -2806,12 +2807,11 @@ static int __get_page_type(struct page_i
             if ( !(x & PGT_partial) )
             {
                 /* Someone else is updating validation of this page. Wait... */
-                while ( (y = page->u.inuse.type_info) == x )
-                {
+                do {
                     if ( preemptible && hypercall_preempt_check() )
                         return -EINTR;
                     cpu_relax();
-                }
+                } while ( (y = page->u.inuse.type_info) == x );
                 continue;
             }
             /* Type ref count was left at 1 when PGT_partial got set. */
